### PR TITLE
add loginform_content plugin hook

### PR DESCRIPTION
**Pull request by johndoh**
implement a new plugin hook `loginform_content` based on discussion in #6569 to allow plugins to change the hidden + visible fields on the login form as well as the buttons.

### CODE REVIEW

File: program/include/rcmail_output_html.php [hunk 1]
```diff
@@ -2356,61 +2375,71 @@ protected function login_form($attrib)
...
+                $form_content['inputs']  = [];
+                $form_content['buttons'] = [];
             }
 
             $link_attr = ['href' => $this->app->url(['action' => 'oauth']), 'id' => 'rcmloginoauth', 'class' => 'button oauth ' . $this->config->get('oauth_provider')];
-            $out .= html::p('oauthlogin', html::a($link_attr, $this->app->gettext(['name' => 'oauthlogin', 'vars' => ['provider' => $this->config->get('oauth_provider_name', 'OAuth')]])));
+            $form_content['buttons']['oauthlogin'] = ['outterclass' => 'oauthlogin' , 'content' => html::a($link_attr, $this->app->gettext(['name' => 'oauthlogin', 'vars' => ['provider' => $this->config->get('oauth_provider_name', 'OAuth')]]))];
+        }
+
+        $data = $this->app->plugins->exec_hook('loginform_content', [
```
Comment (alecpl) [inline program/include/rcmail_output_html.php:2414]: I don't see a use-case for the 'name' argument. And I would just do:
```
$form_content = $this->app->plugins->exec_hook('loginform_content', $form_content);
```

File: program/include/rcmail_output_html.php [hunk 2]
```diff
@@ -2356,61 +2375,71 @@ protected function login_form($attrib)
...
             // hide login form fields when `oauth_login_redirect` is configured
             if ($this->config->get('oauth_login_redirect')) {
-                $out = '';
+                $form_content['inputs']  = [];
+                $form_content['buttons'] = [];
             }
 
             $link_attr = ['href' => $this->app->url(['action' => 'oauth']), 'id' => 'rcmloginoauth', 'class' => 'button oauth ' . $this->config->get('oauth_provider')];
-            $out .= html::p('oauthlogin', html::a($link_attr, $this->app->gettext(['name' => 'oauthlogin', 'vars' => ['provider' => $this->config->get('oauth_provider_name', 'OAuth')]])));
+            $form_content['buttons']['oauthlogin'] = ['outterclass' => 'oauthlogin' , 'content' => html::a($link_attr, $this->app->gettext(['name' => 'oauthlogin', 'vars' => ['provider' => $this->config->get('oauth_provider_name', 'OAuth')]]))];
```
Comment (alecpl) [inline program/include/rcmail_output_html.php:2411]: Too long line.

File: program/include/rcmail_output_html.php [hunk 3]
```diff
@@ -2356,61 +2375,71 @@ protected function login_form($attrib)
...
+            $form_content['buttons']['submit'] = ['outterclass' => 'formbuttons', 'content' => html::tag('button', $button_attr, $this->app->gettext('login'))];
         }
 
         // add oauth login button
         if ($this->config->get('oauth_auth_uri') && $this->config->get('oauth_provider')) {
             // hide login form fields when `oauth_login_redirect` is configured
             if ($this->config->get('oauth_login_redirect')) {
-                $out = '';
+                $form_content['inputs']  = [];
+                $form_content['buttons'] = [];
```
Comment (alecpl) [inline program/include/rcmail_output_html.php:2407]: Here you should clear the 'hidden' element too. No?

File: program/include/rcmail_output_html.php [hunk 4]
```diff
@@ -2356,61 +2375,71 @@ protected function login_form($attrib)
...
+
+        // output login form contents
+        $out = implode('', $data['content']['hidden']);
+
+        if (count($data['content']['inputs']) > 0) {
+            // create HTML table with two cols
+            $table = new html_table(['cols' => 2]);
+
+            foreach ($data['content']['inputs'] as $input) {
+                $table->add('title', $input['title']);
```
Comment (alecpl) [inline program/include/rcmail_output_html.php:2429]: A small improvement: If title is not set put content into a cell with colspan=2.

**Review by alecpl — CHANGES_REQUESTED** [review]

File: program/include/rcmail_output_html.php (hunk 1 above)
Comment (johndoh) [reply program/include/rcmail_output_html.php]: done

File: program/include/rcmail_output_html.php (hunk 3 above)
Comment (johndoh) [reply program/include/rcmail_output_html.php:2408]: done

File: program/include/rcmail_output_html.php (hunk 4 above)
Comment (johndoh) [reply program/include/rcmail_output_html.php]: done

File: program/include/rcmail_output_html.php (hunk 2 above)
Comment (johndoh) [reply program/include/rcmail_output_html.php]: done

**Pull request merged**